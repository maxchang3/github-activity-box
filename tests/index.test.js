import { beforeEach, expect, it, vi } from 'vitest'
import { Issues, PRs } from './fixtures'
const mockedUpdate = vi.fn()
const mockedConsoleError = vi.fn()

vi.spyOn(console, 'log').mockImplementation(() => {})
vi.spyOn(console, 'error').mockImplementation(mockedConsoleError)
vi.spyOn(process, 'exit').mockImplementation(() => {})

vi.mock('gist-box', async (importOriginal) => {
    const actual = await importOriginal()
    return {
        ...actual,
        GistBox: vi.fn().mockImplementation(() => ({
            update: mockedUpdate,
        })),
    }
})

vi.mock('../src/graphql', async () => ({
    getIssues: vi.fn().mockResolvedValue({
        search: {
            edges: Issues,
        },
    }),
    getPRs: vi.fn().mockResolvedValue({
        search: {
            edges: PRs,
        },
    }),
}))

describe('activity-box', () => {
    beforeEach(async () => {
        mockedUpdate.mockClear()
    })

    const runAction = async () => {
        vi.resetModules()
        return await import('../src/index.ts')
    }

    it('updates the Gist with the expected string', async () => {
        await runAction()
        expect(mockedUpdate).toHaveBeenCalledOnce()
        expect(mockedUpdate.mock.calls[0][0]).toMatchSnapshot()
    })

    it('handles failure to update the Gist', async () => {
        mockedUpdate.mockImplementationOnce(() => {
            throw new Error('404')
        })

        await runAction()
        expect(mockedConsoleError).toHaveBeenCalledOnce()
        expect(mockedConsoleError.mock.calls[0][0]).toMatchSnapshot()
        expect(mockedUpdate).toHaveBeenCalled()
    })
})
