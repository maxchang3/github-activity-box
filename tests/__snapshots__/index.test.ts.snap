// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`activity-box > should exclude repos listed in \`EXCLUDE_REPO\` 1`] = `
{
  "content": "❗️ Open: issue #35 in clippy/really-really-really-really-rea...
💪 Open: PR #38 in clippy/take-over-github
❗️ Open: issue #31 in clippy/take-over-github
🎉 Merged: PR #32 in clippy/take-over-github
❌ Closed: PR #30 in clippy/really-really-really-really-reall...",
  "description": undefined,
}
`;

exports[`activity-box > should handle failure to update the Gist 1`] = `"Error getting or update the Gist: Error: 404"`;

exports[`activity-box > should update the Gist with the expected string 1`] = `
{
  "content": "❗️ Open: issue #35 in clippy/really-really-really-really-rea...
💪 Open: PR #50 in clippy/should-be-filtered
💪 Open: PR #38 in clippy/take-over-github
❗️ Open: issue #31 in clippy/take-over-github
🎉 Merged: PR #32 in clippy/take-over-github",
  "description": undefined,
}
`;
